### PR TITLE
Fix insert select deadlock for log family engine

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Bug Fix

### CODE REVIEW

File: dbms/src/Interpreters/InterpreterInsertQuery.cpp [hunk 1]
```diff
@@ -175,4 +128,59 @@ std::pair<String, String> InterpreterInsertQuery::getDatabaseTable() const
...
+        SelectQueryOptions select_query_options = SelectQueryOptions(QueryProcessingStage::Complete, 1);
+        InterpreterSelectWithUnionQuery interpreter_select{query.select, context, select_query_options};
+
+        in = interpreter_select.execute().in;
+        in = std::make_shared<ConvertingBlockInputStream>(context, in, sample_block, ConvertingBlockInputStream::MatchColumnsMode::Position);
+
+        if (!allow_materialized)
+        {
+            Block in_header = in->getHeader();
+            for (const auto &column : table->getColumns())
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/InterpreterInsertQuery.cpp]: Style.

**Comment by alexey-milovidov** [comment]
I cannot instantly figure out from the diff, what exact changes were made.

**Comment by zhang2014** [comment]
> I cannot instantly figure out from the diff, what exact changes were made.

This fix is a bit of a hack, avoiding deadlock by prioritizing insert and select execution, and I've added some comments to the code.

https://github.com/yandex/ClickHouse/pull/5097/files#diff-0a9ad6b0fe82092d9bf054174f8bac51R102

**Comment by tavplubix** [comment]
Will be assigned to @Akazz

**Comment by Akazz** [comment]
- ClickHouse does not provide "snapshot isolation" for queries  - only a simple "multiple readers/single writer" model (reads and writes cannot run on the same data concurrently)
 - In the absence of such isolation the semantics of `INSERT INTO $self SELECT * from $self` is unclear
 - Technically `INSERT INTO $self SELECT * from $self` queries would lock up for most of the ENGINEs
 - Even though StripeLog and Log engines do not lock up with your patch their are not straightforward and might fail in some scenarios
 - Additionally, now with @akuzm's #6226 PR having been merged your patch leads to another deadlock for TinyLog engine
 
 **Summary**: it is unclear at the moment if this PR is worth playing with.

**Comment by Akazz** [comment]
@zhang2014

> ClickHouse does not provide "snapshot isolation" for queries

This above is too strong a statement and of course we have important exceptions to it: CH provides snapshot isolation for *MergeTree as well as for Memory, Log/StripeLog engines (due to the fact that new data is only being appended).

A few changes to `StorageLog::loadMarks()` mechanism might be useful as well (blocking is only needed while the marks are being loaded for the first time).

File: dbms/src/Interpreters/InterpreterInsertQuery.h [hunk 2]
```diff
@@ -34,6 +34,12 @@ class InterpreterInsertQuery : public IInterpreter
     ASTPtr query_ptr;
     const Context & context;
     bool allow_materialized;
+
+
+    BlockInputStreamPtr tryCreateSourceInputStream(const ASTInsertQuery & query, const StoragePtr & table, const Block & sample_block);
```
Comment (Akazz) [inline dbms/src/Interpreters/InterpreterInsertQuery.h]: It might be a good idea to pass StoragePtr by value

**Comment by zhang2014** [comment]
Continued by @Akazz  in #6802

**Review by 4ertus2 — COMMENTED** [review]
.

**Comment by alexey-milovidov** [comment]
Congratulations, it's merged!

**Pull request closed without merging**